File: examples/android/retail-wallet/src/main/java/org/hyperledger/iroha/samples/wallet/PosManifest.kt
package org.hyperledger.iroha.samples.wallet

import android.content.Context
import java.io.BufferedReader
import java.io.InputStreamReader
import java.security.KeyFactory
import java.security.Signature
import java.security.spec.X509EncodedKeySpec
import java.time.Clock
import java.time.Duration
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.Locale
import java.util.Base64
import org.json.JSONArray
import org.json.JSONObject
import org.hyperledger.iroha.android.address.AccountAddress
import org.hyperledger.iroha.android.address.AccountAddressException

data class PosProvisionManifest(
    val manifestId: String,
    val sequence: Int,
    val publishedAtMs: Long,
    val validFromMs: Long,
    val validUntilMs: Long,
    val rotationHintMs: Long?,
    val operator: String,
    val backendRoots: List<PosBackendRoot>,
    val metadata: JSONObject?,
    val payloadBase64: String,
    val operatorSignature: String
)

data class PosBackendRoot(
    val label: String,
    val role: String,
    val publicKey: String,
    val validFromMs: Long,
    val validUntilMs: Long,
    val metadata: JSONObject?
)

data class ManifestStatus(
    val manifestId: String,
    val sequence: Int,
    val operator: String,
    val validWindowLabel: String,
    val rotationLabel: String,
    val dualStatusLabel: String,
    val dualStatusHealthy: Boolean,
    val warnings: List<String>,
    val backendRoots: List<BackendRootStatus>
) {
    companion object {
        fun from(
            manifest: PosProvisionManifest,
            clock: Clock = Clock.systemUTC()
        ): ManifestStatus {
            val formatter = DateTimeFormatter.ISO_INSTANT.withZone(ZoneOffset.UTC)
            val validLabel = formatter.format(Instant.ofEpochMilli(manifest.validFromMs)) +
                " – " + formatter.format(Instant.ofEpochMilli(manifest.validUntilMs))
            val rotationLabel = manifest.rotationHintMs?.let { hint ->
                formatter.format(Instant.ofEpochMilli(hint))
            } ?: "n/a"
            val now = clock.millis()
            val backendStatuses = manifest.backendRoots.map { root ->
                val active = now in root.validFromMs..root.validUntilMs
                val statusLabel = buildBackendStatusLabel(root, active, formatter, now)
                BackendRootStatus(
                    label = root.label,
                    role = root.role,
                    statusLabel = statusLabel,
                    active = active
                )
            }
            val dualStatus = buildDualStatus(manifest, backendStatuses, now)
            val warnings = buildWarnings(manifest, backendStatuses, dualStatus, now, formatter)
            return ManifestStatus(
                manifestId = manifest.manifestId,
                sequence = manifest.sequence,
                operator = manifest.operator,
                validWindowLabel = validLabel,
                rotationLabel = rotationLabel,
                dualStatusLabel = dualStatus.label,
                dualStatusHealthy = dualStatus.healthy,
                warnings = warnings,
                backendRoots = backendStatuses
            )
        }

        private fun buildBackendStatusLabel(
            root: PosBackendRoot,
            active: Boolean,
            formatter: DateTimeFormatter,
            nowMs: Long
        ): String {
            val expiresIn = root.validUntilMs - nowMs
            val expiresLabel = if (expiresIn > 0) {
                "expires ${formatDuration(expiresIn)}"
            } else {
                "expired ${formatter.format(Instant.ofEpochMilli(root.validUntilMs))}"
            }
            val validity = formatter.format(Instant.ofEpochMilli(root.validFromMs)) +
                " – " + formatter.format(Instant.ofEpochMilli(root.validUntilMs))
            val state = if (active) "active" else "inactive"
            return "$state · $validity ($expiresLabel)"
        }

        private fun buildDualStatus(
            manifest: PosProvisionManifest,
            roots: List<BackendRootStatus>,
            now: Long
        ): DualStatus {
            val roles = roots.filter { it.active }.map { it.role }.toSet()
            val required = setOf("offline_admission_signer", "offline_allowance_witness")
            val missing = required.filterNot { roles.contains(it) }
            return if (missing.isEmpty()) {
                DualStatus(true, "admission + witness present until ${formatDuration(manifest.validUntilMs - now)}")
            } else {
                DualStatus(false, "missing ${missing.joinToString(", ")}")
            }
        }

        private fun buildWarnings(
            manifest: PosProvisionManifest,
            roots: List<BackendRootStatus>,
            dualStatus: DualStatus,
            now: Long,
            formatter: DateTimeFormatter
        ): List<String> {
            val warnings = mutableListOf<String>()
            if (now < manifest.validFromMs) {
                warnings.add("manifest not active until ${formatter.format(Instant.ofEpochMilli(manifest.validFromMs))}")
            }
            val manifestExpiresIn = manifest.validUntilMs - now
            if (manifestExpiresIn <= MANIFEST_WARNING_WINDOW_MS) {
                warnings.add("manifest expires in ${formatDuration(manifestExpiresIn)}")
            }
            manifest.rotationHintMs?.let { hint ->
                if (now >= hint) {
                    warnings.add("rotation hint passed ${formatDuration(now - hint)} ago")
                } else if (hint - now <= ROTATION_WARNING_WINDOW_MS) {
                    warnings.add("rotation hint in ${formatDuration(hint - now)}")
                }
            }
            roots.filter { it.active.not() }.forEach { inactive ->
                warnings.add("${inactive.label} inactive")
            }
            if (!dualStatus.healthy) {
                warnings.add(dualStatus.label)
            }
            if (warnings.isEmpty()) {
                return emptyList()
            }
            return warnings
        }

        private fun formatDuration(durationMs: Long): String {
            if (durationMs <= 0) {
                return "0s"
            }
            val duration = Duration.ofMillis(durationMs)
            val days = duration.toDays()
            val hours = duration.minusDays(days).toHours()
            val minutes = duration.minusDays(days).minusHours(hours).toMinutes()
            val parts = mutableListOf<String>()
            if (days > 0) {
                parts.add("${days}d")
            }
            if (hours > 0) {
                parts.add("${hours}h")
            }
            if (minutes > 0) {
                parts.add("${minutes}m")
            }
            if (parts.isEmpty()) {
                parts.add("${duration.seconds}s")
            }
            return parts.joinToString(" ")
        }
    }
}

data class BackendRootStatus(
    val label: String,
    val role: String,
    val statusLabel: String,
    val active: Boolean
)

private data class DualStatus(val healthy: Boolean, val label: String)

object PosManifestLoader {
    private const val MANIFEST_ASSET = "pos_manifest.json"

    fun loadFromAssets(context: Context): PosProvisionManifest {
        context.assets.open(MANIFEST_ASSET).use { input ->
            val reader = BufferedReader(InputStreamReader(input))
            val raw = reader.readText()
            return parse(raw)
        }
    }

    fun parse(raw: String): PosProvisionManifest {
        val json = JSONObject(raw)
        val backendRoots = parseBackendRoots(json.optJSONArray("backend_roots"))
        val payloadBase64 =
            json.optString("payload_base64")
                .takeIf { it.isNotBlank() }
                ?: throw IllegalArgumentException("manifest missing payload_base64")
        val operatorSignature =
            json.optString("operator_signature")
                .takeIf { it.isNotBlank() }
                ?: throw IllegalArgumentException("manifest missing operator_signature")
        val operator = json.getString("operator")
        val payload = decodeBase64(payloadBase64)
        val signature = decodeHex(operatorSignature)
        val operatorKey = decodeOperatorPublicKey(operator)
        verifySignature(operatorKey, payload, signature)
        return PosProvisionManifest(
            manifestId = json.getString("manifest_id"),
            sequence = json.getInt("sequence"),
            publishedAtMs = json.getLong("published_at_ms"),
            validFromMs = json.getLong("valid_from_ms"),
            validUntilMs = json.getLong("valid_until_ms"),
            rotationHintMs = json.optLong("rotation_hint_ms").takeIf { json.has("rotation_hint_ms") },
            operator = operator,
            backendRoots = backendRoots,
            metadata = json.optJSONObject("metadata"),
            payloadBase64 = payloadBase64,
            operatorSignature = operatorSignature
        )
    }

    private fun parseBackendRoots(array: JSONArray?): List<PosBackendRoot> {
        if (array == null) return emptyList()
        val items = mutableListOf<PosBackendRoot>()
        for (index in 0 until array.length()) {
            val entry = array.getJSONObject(index)
            items.add(
                PosBackendRoot(
                    label = entry.getString("label"),
                    role = entry.getString("role").lowercase(Locale.US),
                    publicKey = entry.getString("public_key"),
                    validFromMs = entry.getLong("valid_from_ms"),
                    validUntilMs = entry.getLong("valid_until_ms"),
                    metadata = entry.optJSONObject("metadata")
                )
            )
        }
        return items
    }

    private fun decodeOperatorPublicKey(operatorId: String): ByteArray {
        val address = operatorId.substringBefore("@")
        if (address.lowercase(Locale.US).startsWith("ed01")) {
            val raw = decodeHex(address)
            require(raw.size > 3 && raw[0] == 0xED.toByte() && raw[1] == 0x01.toByte()) {
                "unsupported operator key encoding"
            }
            val declaredLen = raw[2].toInt() and 0xFF
            require(declaredLen == raw.size - 3) { "unexpected operator key length" }
            return raw.copyOfRange(3, raw.size)
        }
        try {
            val decoded = AccountAddress.fromIH58(address, null)
            return extractSingleSignatoryKey(decoded.canonicalBytes())
        } catch (ex: AccountAddressException) {
            throw IllegalArgumentException("failed to parse operator account", ex)
        }
    }

    private fun extractSingleSignatoryKey(canonical: ByteArray): ByteArray {
        if (canonical.size < 4) {
            throw IllegalArgumentException("invalid canonical address length")
        }
        var cursor = 0
        val header = canonical[cursor++].toInt() and 0xFF
        val extensionFlag = header and 0x01
        val classBits = (header shr 3) and 0x03
        require(extensionFlag == 0) { "address extension flag set" }
        require(classBits == 0 || classBits == 1) { "unknown address class" }
        cursor += when (val domainTag = canonical[cursor++].toInt() and 0xFF) {
            0x00 -> 0
            0x01 -> 12
            0x02 -> 4
            else -> throw IllegalArgumentException("unknown domain tag: $domainTag")
        }
        if (cursor >= canonical.size) {
            throw IllegalArgumentException("invalid canonical address length")
        }
        val controllerTag = canonical[cursor++].toInt() and 0xFF
        require(controllerTag == 0x00) { "unsupported controller tag $controllerTag" }
        if (cursor + 2 > canonical.size) {
            throw IllegalArgumentException("invalid canonical address length")
        }
        val curveId = canonical[cursor++].toInt() and 0xFF
        require(curveId == 0x01) { "unsupported signing algorithm $curveId" }
        val keyLen = canonical[cursor++].toInt() and 0xFF
        val end = cursor + keyLen
        if (end != canonical.size) {
            throw IllegalArgumentException("unexpected trailing bytes in address payload")
        }
        return canonical.copyOfRange(cursor, end)
    }

    private fun decodeBase64(value: String): ByteArray {
        return try {
            Base64.getDecoder().decode(value)
        } catch (ex: IllegalArgumentException) {
            throw IllegalArgumentException("invalid base64 payload", ex)
        }
    }

    private fun decodeHex(value: String): ByteArray {
        val normalized = value.trim()
        if (normalized.length % 2 != 0) {
            throw IllegalArgumentException("hex string must have even length")
        }
        val out = ByteArray(normalized.length / 2)
        var index = 0
        while (index < normalized.length) {
            val hi = Character.digit(normalized[index], 16)
            val lo = Character.digit(normalized[index + 1], 16)
            if (hi == -1 || lo == -1) {
                throw IllegalArgumentException("invalid hex string")
            }
            out[index / 2] = ((hi shl 4) or lo).toByte()
            index += 2
        }
        return out
    }

    private fun verifySignature(publicKey: ByteArray, payload: ByteArray, signature: ByteArray) {
        require(publicKey.size == 32) { "operator public key must be 32 bytes" }
        val encoded = encodeEd25519PublicKey(publicKey)
        try {
            val keyFactory = KeyFactory.getInstance("Ed25519")
            val publicSpec = X509EncodedKeySpec(encoded)
            val verifier = Signature.getInstance("Ed25519")
            verifier.initVerify(keyFactory.generatePublic(publicSpec))
            verifier.update(payload)
            if (!verifier.verify(signature)) {
                throw IllegalArgumentException("manifest signature verification failed")
            }
        } catch (ex: Exception) {
            if (ex is IllegalArgumentException) throw ex
            throw IllegalArgumentException("manifest signature verification failed", ex)
        }
    }

    private fun encodeEd25519PublicKey(raw: ByteArray): ByteArray {
        // RFC 8410 SubjectPublicKeyInfo prefix for Ed25519
        val prefix =
            byteArrayOf(
                0x30, 0x2A,
                0x30, 0x05,
                0x06, 0x03,
                0x2B, 0x65, 0x70,
                0x03, 0x21, 0x00
            )
        return ByteArray(prefix.size + raw.size).also {
            System.arraycopy(prefix, 0, it, 0, prefix.size)
            System.arraycopy(raw, 0, it, prefix.size, raw.size)
        }
    }
}

private const val MANIFEST_WARNING_WINDOW_MS = 7L * 24 * 60 * 60 * 1000 // 7 days
private const val ROTATION_WARNING_WINDOW_MS = 3L * 24 * 60 * 60 * 1000 // 3 days

File: examples/ios/NoritoDemo/Sources/PosManifest.swift
import Foundation
import CryptoKit
#if canImport(IrohaSwift)
import IrohaSwift
#endif

struct PosProvisionManifest: Decodable {
  let manifestId: String
  let sequence: Int
  let publishedAtMs: UInt64
  let validFromMs: UInt64
  let validUntilMs: UInt64
  let rotationHintMs: UInt64?
  let operatorId: String
  let backendRoots: [PosBackendRoot]
  let metadata: [String: String]?
  let payloadBase64: String
  let operatorSignature: String?

  enum CodingKeys: String, CodingKey {
    case manifestId = "manifest_id"
    case sequence
    case publishedAtMs = "published_at_ms"
    case validFromMs = "valid_from_ms"
    case validUntilMs = "valid_until_ms"
    case rotationHintMs = "rotation_hint_ms"
    case operatorId = "operator"
    case backendRoots = "backend_roots"
    case metadata
    case payloadBase64 = "payload_base64"
    case operatorSignature = "operator_signature"
  }
}

struct PosBackendRoot: Decodable {
  let label: String
  let role: String
  let publicKey: String
  let validFromMs: UInt64
  let validUntilMs: UInt64
  let metadata: [String: String]?

  enum CodingKeys: String, CodingKey {
    case label
    case role
    case publicKey = "public_key"
    case validFromMs = "valid_from_ms"
    case validUntilMs = "valid_until_ms"
    case metadata
  }
}

struct PosManifestStatus: Identifiable {
  struct BackendRootStatus: Identifiable {
    var id: String { label }
    let label: String
    let role: String
    let statusLabel: String
    let active: Bool
  }

  var id: String { manifestId }
  let manifestId: String
  let sequence: Int
  let operatorId: String
  let validWindowLabel: String
  let rotationLabel: String
  let dualStatusLabel: String
  let dualStatusHealthy: Bool
  let warnings: [String]
  let backendRoots: [BackendRootStatus]

  static func from(manifest: PosProvisionManifest, now: Date = Date()) -> PosManifestStatus {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    let validWindow = "\(formatter.string(from: manifest.validFromMs.date)) – \(formatter.string(from: manifest.validUntilMs.date))"
    let rotationLabel = manifest.rotationHintMs.map { formatter.string(from: $0.date) } ?? "n/a"
    let backendStatuses = manifest.backendRoots.map { root -> BackendRootStatus in
      let active = now >= root.validFromMs.date && now <= root.validUntilMs.date
      let statusText = backendStatusLabel(root: root, active: active, formatter: formatter, now: now)
      return BackendRootStatus(label: root.label, role: root.role, statusLabel: statusText, active: active)
    }
    let dualStatus = computeDualStatus(roots: backendStatuses, manifest: manifest, now: now)
    let warnings = computeWarnings(manifest: manifest, roots: backendStatuses, dualStatus: dualStatus, now: now, formatter: formatter)
    return PosManifestStatus(
      manifestId: manifest.manifestId,
      sequence: manifest.sequence,
      operatorId: manifest.operatorId,
      validWindowLabel: validWindow,
      rotationLabel: rotationLabel,
      dualStatusLabel: dualStatus.label,
      dualStatusHealthy: dualStatus.healthy,
      warnings: warnings,
      backendRoots: backendStatuses
    )
  }

  static func unavailable(message: String) -> PosManifestStatus {
    return PosManifestStatus(
      manifestId: "unavailable",
      sequence: -1,
      operatorId: "n/a",
      validWindowLabel: "n/a",
      rotationLabel: "n/a",
      dualStatusLabel: message,
      dualStatusHealthy: false,
      warnings: [message],
      backendRoots: []
    )
  }

  private struct DualStatus {
    let healthy: Bool
    let label: String
  }

  private static func backendStatusLabel(root: PosBackendRoot, active: Bool, formatter: ISO8601DateFormatter, now: Date) -> String {
    let expiresIn = root.validUntilMs.date.timeIntervalSince(now)
    let expiresLabel: String
    if expiresIn > 0 {
      expiresLabel = "expires \(formatDuration(milliseconds: expiresIn * 1000))"
    } else {
      expiresLabel = "expired \(formatter.string(from: root.validUntilMs.date))"
    }
    let window = "\(formatter.string(from: root.validFromMs.date)) – \(formatter.string(from: root.validUntilMs.date))"
    return "\(active ? "active" : "inactive") · \(window) (\(expiresLabel))"
  }

  private static func computeDualStatus(roots: [BackendRootStatus], manifest: PosProvisionManifest, now: Date) -> DualStatus {
    let activeRoles = Set(roots.filter(\.active).map { $0.role })
    let required: Set<String> = ["offline_admission_signer", "offline_allowance_witness"]
    let missing = required.subtracting(activeRoles)
    if missing.isEmpty {
      let remaining = manifest.validUntilMs.date.timeIntervalSince(now)
      return DualStatus(healthy: true, label: "admission+witness for \(formatDuration(milliseconds: remaining * 1000))")
    }
    return DualStatus(healthy: false, label: "missing \(missing.joined(separator: \", \"))")
  }

  private static func computeWarnings(
    manifest: PosProvisionManifest,
    roots: [BackendRootStatus],
    dualStatus: DualStatus,
    now: Date,
    formatter: ISO8601DateFormatter
  ) -> [String] {
    var warnings: [String] = []
    if now < manifest.validFromMs.date {
      warnings.append("manifest not active until \(formatter.string(from: manifest.validFromMs.date))")
    }
    let manifestExpires = manifest.validUntilMs.date.timeIntervalSince(now)
    if manifestExpires <= PosManifestLoader.manifestWarningWindow {
      warnings.append("manifest expires in \(formatDuration(milliseconds: manifestExpires * 1000))")
    }
    if let hint = manifest.rotationHintMs {
      let delta = hint.date.timeIntervalSince(now)
      if delta <= 0 {
        warnings.append("rotation hint passed \(formatDuration(milliseconds: -delta * 1000)) ago")
      } else if delta <= PosManifestLoader.rotationWarningWindow {
        warnings.append("rotation hint in \(formatDuration(milliseconds: delta * 1000))")
      }
    }
    roots.filter { !$0.active }.forEach { root in
      warnings.append("\(root.label) inactive")
    }
    if !dualStatus.healthy {
      warnings.append(dualStatus.label)
    }
    return warnings
  }

  private static func formatDuration(milliseconds: Double) -> String {
    if milliseconds <= 0 {
      return "0s"
    }
    var seconds = Int(milliseconds / 1000)
    let days = seconds / (24 * 3600)
    seconds %= 24 * 3600
    let hours = seconds / 3600
    seconds %= 3600
    let minutes = seconds / 60
    let parts = [
      days > 0 ? "\(days)d" : nil,
      hours > 0 ? "\(hours)h" : nil,
      minutes > 0 ? "\(minutes)m" : nil
    ].compactMap { $0 }
    if parts.isEmpty {
      return "\(Int(milliseconds / 1000))s"
    }
    return parts.joined(separator: " ")
  }
}

enum PosManifestLoader {
  static let manifestWarningWindow: TimeInterval = 7 * 24 * 3600
  static let rotationWarningWindow: TimeInterval = 3 * 24 * 3600

  static func loadManifest(bundle: Bundle = .main) throws -> PosProvisionManifest {
    guard let url = bundle.url(forResource: "pos_manifest", withExtension: "json") else {
      throw NSError(domain: "PosManifestLoader", code: 1, userInfo: [NSLocalizedDescriptionKey: "pos_manifest.json missing from bundle"])
    }
    let data = try Data(contentsOf: url)
    return try parse(data: data)
  }

  static func parse(data: Data) throws -> PosProvisionManifest {
    let decoder = JSONDecoder()
    decoder.keyDecodingStrategy = .useDefaultKeys
    let manifest = try decoder.decode(PosProvisionManifest.self, from: data)
    try verifySignature(manifest: manifest)
    return manifest
  }

  private static func verifySignature(manifest: PosProvisionManifest) throws {
    guard let operatorSignature = manifest.operatorSignature,
          let signature = Data(hexString: operatorSignature) else {
      throw NSError(
        domain: "PosManifestLoader",
        code: 2,
        userInfo: [NSLocalizedDescriptionKey: "manifest missing operator signature"]
      )
    }
    guard let payload = Data(base64Encoded: manifest.payloadBase64) else {
      throw NSError(
        domain: "PosManifestLoader",
        code: 3,
        userInfo: [NSLocalizedDescriptionKey: "manifest payload_base64 invalid"]
      )
    }
    let operatorKey = try decodeOperatorPublicKey(manifest.operatorId)
    let publicKey = try Curve25519.Signing.PublicKey(rawRepresentation: operatorKey)
    guard publicKey.isValidSignature(signature, for: payload) else {
      throw NSError(
        domain: "PosManifestLoader",
        code: 4,
        userInfo: [NSLocalizedDescriptionKey: "manifest signature verification failed"]
      )
    }
  }

  private static func decodeOperatorPublicKey(_ operatorId: String) throws -> Data {
    let address = operatorId.split(separator: "@", maxSplits: 1, omittingEmptySubsequences: false).first.map(String.init) ?? operatorId
    let normalized = address.trimmingCharacters(in: .whitespacesAndNewlines)
    if normalized.lowercased().hasPrefix("ed01") {
      guard let raw = Data(hexString: normalized) else {
        throw NSError(domain: "PosManifestLoader", code: 5, userInfo: [NSLocalizedDescriptionKey: "invalid operator key encoding"])
      }
      guard raw.count > 3, raw[0] == 0xED, raw[1] == 0x01 else {
        throw NSError(domain: "PosManifestLoader", code: 6, userInfo: [NSLocalizedDescriptionKey: "unsupported operator key encoding"])
      }
      let declaredLen = Int(raw[2])
      guard declaredLen == raw.count - 3 else {
        throw NSError(domain: "PosManifestLoader", code: 7, userInfo: [NSLocalizedDescriptionKey: "unexpected operator key length"])
      }
      return Data(raw.dropFirst(3))
    }
#if canImport(IrohaSwift)
    let account = try AccountAddress.fromI105(normalized, expectedPrefix: nil)
    let canonical = try account.canonicalBytes()
    return try extractSingleSignatoryKey(canonical)
#else
    throw NSError(domain: "PosManifestLoader", code: 8, userInfo: [NSLocalizedDescriptionKey: "I105 decoding requires IrohaSwift"])
#endif
  }

#if canImport(IrohaSwift)
  private static func extractSingleSignatoryKey(_ canonical: Data) throws -> Data {
    guard canonical.count >= 4 else {
      throw NSError(domain: "PosManifestLoader", code: 9, userInfo: [NSLocalizedDescriptionKey: "invalid canonical address length"])
    }
    var cursor = canonical.startIndex
    let header = canonical[cursor]
    cursor += 1
    let extensionFlag = header & 0x01
    let classBits = (header >> 3) & 0x03
    guard extensionFlag == 0 else {
      throw NSError(domain: "PosManifestLoader", code: 10, userInfo: [NSLocalizedDescriptionKey: "address extension flag set"])
    }
    guard classBits == 0 || classBits == 1 else {
      throw NSError(domain: "PosManifestLoader", code: 11, userInfo: [NSLocalizedDescriptionKey: "unknown address class"])
    }
    guard cursor < canonical.endIndex else {
      throw NSError(domain: "PosManifestLoader", code: 12, userInfo: [NSLocalizedDescriptionKey: "invalid canonical address length"])
    }
    let domainTag = canonical[cursor]
    cursor += 1
    switch domainTag {
    case 0x00:
      break
    case 0x01:
      cursor += 12
    case 0x02:
      cursor += 4
    default:
      throw NSError(domain: "PosManifestLoader", code: 13, userInfo: [NSLocalizedDescriptionKey: "unknown domain tag \(domainTag)"])
    }
    guard cursor < canonical.endIndex else {
      throw NSError(domain: "PosManifestLoader", code: 14, userInfo: [NSLocalizedDescriptionKey: "invalid canonical address length"])
    }
    let controllerTag = canonical[cursor]
    cursor += 1
    guard controllerTag == 0x00 else {
      throw NSError(domain: "PosManifestLoader", code: 15, userInfo: [NSLocalizedDescriptionKey: "unsupported controller tag \(controllerTag)"])
    }
    guard cursor + 2 <= canonical.endIndex else {
      throw NSError(domain: "PosManifestLoader", code: 16, userInfo: [NSLocalizedDescriptionKey: "invalid canonical address length"])
    }
    let curveId = canonical[cursor]
    cursor += 1
    guard curveId == 0x01 else {
      throw NSError(domain: "PosManifestLoader", code: 17, userInfo: [NSLocalizedDescriptionKey: "unsupported signing algorithm \(curveId)"])
    }
    let keyLen = Int(canonical[cursor])
    cursor += 1
    let end = cursor + keyLen
    guard end == canonical.count else {
      throw NSError(domain: "PosManifestLoader", code: 18, userInfo: [NSLocalizedDescriptionKey: "unexpected trailing bytes in address payload"])
    }
    return Data(canonical[cursor..<end])
  }
#endif
}
}

private extension UInt64 {
  var date: Date { Date(timeIntervalSince1970: TimeInterval(self) / 1000.0) }
}
